lib/cc-cubic.c

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (c) 2019 Fastly, Janardhan Iyengar
+ * Copyright (c) 2020 RWTH Aachen University, COMSYS Network Architectures Group, Leo Blöcher
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include <math.h>
+#include "quicly/cc.h"
+
+#define QUICLY_MIN_CWND 2
+
+typedef double cubic_float_t;
```

Comment:
One of the things still on my todo list is comparing this implementation to other QUIC stacks with CUBIC (mvfst, msquic) and maybe TCP stacks to see how they implement the arithmetic operations.

Storing/calculating everything as doubles probably isn't the fastest way, but the one implementation I investigated prior to starting work on this ([quiche](https://github.com/cloudflare/quiche/blob/master/src/recovery/cubic.rs)) also used 64bit floats. Open for suggestions!

Reply:
I'm fine with starting with a double. I'd assume that most platforms capable of running crypto at certain speed would not suffer from us doing FP in CC.

Reply:
It appears that msquic is the only implementation out of the ones I mentioned that uses integer arithmetic for CUBIC. mvfst has TODOs for benchmarking different cbrt implementations against each other, so maybe their decision will provide some insight. After putting some more thought into this I don't think the effort is currently worth it.